### PR TITLE
Support more wallets with Onboard.js

### PR DESCRIPTION
**Pull request by chrismaree**
This PR adds Onboard.j. These add a _much_ wider range of web3 connectors and gives transaction tracking. The current config works on mainnet & Kovan.

Wallet connector on desktop 
image

Wallet connector on mobile
image

Pending transactions on main net:
image
image

Note that in the current implementation Notify.js will break if running on kovan. It detects the transaction has been sent but does not proceed to notify when the transaction has `succeeded`. Other than the network switching on notify.js this implementation works perfectly between networks.

I have validated the behavior of this PR by making a main net position and doubled checked the correct notifications were generated.

Close: #82

### CODE REVIEW

File: features/manage-position/Withdraw.tsx [hunk 1]
```diff
@@ -440,4 +445,4 @@ const Deposit = () => {
   }
 };
 
-export default Deposit;
+export default Withdraw;
```
Comment (chrismaree) [inline features/manage-position/Withdraw.tsx:448]: unrelated to this PR but this was a typo we've had for a while

File: containers/Connection.ts [hunk 2]
```diff
@@ -1,54 +1,116 @@
...
   const [address, setAddress] = useState<string | null>(null);
+  const [selectedWallet, setSelectedWallet] = useState<string | null>(null);
   const [error, setError] = useState<Error | null>(null);
   const [block$, setBlock$] = useState<Observable<Block> | null>(null);
 
   const attemptConnection = async () => {
-    if (window.ethereum === undefined) {
-      throw Error("MetaMask not found, please visit https://metamask.io/");
-    }
+    const apiKey = process.env.REACT_APP_ONBOARD_API_KEY
```
Comment (chrismaree) [inline containers/Connection.ts:28]: note the dapp will work fine without specifying any of these params. They default to common public values (public Infura key & the default onboard.js API key)

File: containers/Collateral.ts [hunk 3]
```diff
@@ -51,7 +51,11 @@ function useCollateral() {
   const setMaxAllowance = async () => {
     if (contract && empAddress) {
       try {
-        await contract.approve(empAddress, ethers.constants.MaxUint256);
+        const tx = await contract.approve(
+          empAddress,
+          ethers.constants.MaxUint256
+        );
+        notify?.hash(tx.hash);
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply containers/Collateral.ts:58]: I think it sends a [notification](https://docs.blocknative.com/notify)

File: containers/Connection.ts (hunk 2 above)
Comment (nicholaspai) [reply containers/Connection.ts:28]: Going off matt's questions, who's keys are these and why would you ever want to set your own keys?

File: containers/Connection.ts [hunk 4]
```diff
@@ -1,54 +1,112 @@
...
+            walletName: "imToken",
+            rpcUrl:
+              !!network && network.chainId === 1
+                ? "https://mainnet-eth.token.im"
+                : "https://eth-testnet.tokenlon.im",
+            preferred: true,
+          },
+          { walletName: "coinbase", preferred: true },
+          {
+            walletName: "portis",
```
Comment (nicholaspai) [inline containers/Connection.ts:72]: When I tried to test this out I only got Metamask option

File: containers/Connection.ts [hunk 5]
```diff
@@ -1,54 +1,112 @@
...
+            const ethersProvider = new ethers.providers.Web3Provider(
+              wallet.provider
+            );
+            setProvider(ethersProvider);
+            setNetwork(await ethersProvider.getNetwork());
+            setSigner(ethersProvider.getSigner());
+            await ethersProvider.send("eth_requestAccounts", []);
+            setAddress(await ethersProvider.getSigner().getAddress());
+            setSelectedWallet(wallet.name);
+          } else {
```
Comment (nicholaspai) [inline containers/Connection.ts:54]: I think you should set the Provider and all of these values to null together. Is it possible for network, signer, address, etc. to be non null if provider `wallet.provider` doesn't exist?

File: features/manage-position/Create.tsx [hunk 6]
```diff
@@ -143,6 +145,7 @@ const Create = () => {
           const tokensWei = toWei(tokens);
           const tx = await emp.create([collateralWei], [tokensWei]);
           setHash(tx.hash as string);
+          notify?.hash(tx.hash);
```
Comment (nicholaspai) [inline features/manage-position/Create.tsx:148]: Should we remove the the transaction hash/success/error component from this and other files now?

File: features/manage-position/Withdraw.tsx (hunk 1 above)
Comment (nicholaspai) [reply features/manage-position/Withdraw.tsx:448]: Lol woops

**Comment by adrianmcli** [comment]
This looks like a great UX improvement. I'm not familiar w/ the Onboard API, but it seems like things are laid out alright.

File: containers/Connection.ts (hunk 2 above)
Comment (chrismaree) [reply containers/Connection.ts:28]: This is the sample key from the onboard.js repo. note that not all wallets need this param. Metamask, for example, does not require you to provide this key. 

I'm not actually sure how the dos protection would work in this case. I mean if it's set as a .env variable for the server then the risk is pretty low as no one can actually find it and dos it. In the official docs it's just that some wallet providers require this as a field 🤷‍♂️ https://docs.blocknative.com/onboard#built-in-modules
I can reach out to them and ask about this specific point as I'm not actually sure.

File: containers/Collateral.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrismaree) [reply containers/Collateral.ts:58]: yup. wherever we generate a hash you pass it to the `notify` object and it will track it until it enters a block and provides notifications to the user.

File: containers/Connection.ts (hunk 5 above)
Comment (chrismaree) [reply containers/Connection.ts:54]: done with setting everything else to null.

I'm not sure about how it handles edge cases. need to test that a bit more.

File: containers/Connection.ts (hunk 4 above)
Comment (chrismaree) [reply containers/Connection.ts:76]: really? can you post a screenshot? My side looks like this on desktop:
image

File: containers/Connection.ts [hunk 7]
```diff
@@ -1,54 +1,112 @@
...
+            walletName: "walletConnect",
+            rpc: { [network?.chainId || 1]: infuraRpc },
+          },
+          { walletName: "walletLink", rpcUrl: infuraRpc },
+          { walletName: "opera" },
+          { walletName: "operaTouch" },
+          { walletName: "torus" },
+          { walletName: "status" },
+          { walletName: "unilogin" },
+          // { walletName: "authereum" },
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply containers/Connection.ts:91]: I've seen reports that it can be a bit dodgy so removed it for now.

File: features/manage-position/Create.tsx (hunk 6 above)
Comment (chrismaree) [reply features/manage-position/Create.tsx:148]: I thought about that. I dont really see any reason to leaving them both for now. Can strip them out if need be but it's useful to get both I think

File: containers/Collateral.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicholaspai) [reply containers/Collateral.ts:58]: What happens if there are multiple transactions pending at the same time for the user?

File: containers/Connection.ts (hunk 2 above)
Comment (nicholaspai) [reply containers/Connection.ts:28]: Is there any way to avoid having these env vars in the dApp? They're not very clean and I'd like to avoid having these .env dependencies even if they're not a problem in run-time.

File: containers/Connection.ts [hunk 8]
```diff
@@ -1,54 +1,116 @@
...
+      },
+      walletSelect: {
+        wallets: [
+          { walletName: "metamask", preferred: true },
+          {
+            walletName: "imToken",
+            rpcUrl:
+              !!network && network.chainId === 1
+                ? "https://mainnet-eth.token.im"
+                : "https://eth-testnet.tokenlon.im",
```
Comment (nicholaspai) [inline containers/Connection.ts:71]: Can you explain this? "imToken" uses two servers, one for mainnet and one for everything else?

File: containers/Connection.ts [hunk 9]
```diff
@@ -1,54 +1,116 @@
...
+            preferred: true,
+          },
+          { walletName: "coinbase", preferred: true },
+          {
+            walletName: "portis",
+            apiKey: process.env.REACT_APP_PORTIS_API_KEY,
+          },
+          { walletName: "trust", rpcUrl: infuraRpc },
```
Comment (nicholaspai) [inline containers/Connection.ts:79]: So TrustWallet, Dapper, WalletLink, all require an infura node to work? I thought they would have their own provider to iinject?

File: features/weth/Weth.tsx [hunk 10]
```diff
@@ -81,7 +81,7 @@ const Weth = () => {
     wethBalance,
     ethBalance,
   } = WethContract.useContainer();
-  const { signer } = Connection.useContainer();
+  const { signer, notify } = Connection.useContainer();
```
Comment (nicholaspai) [inline features/weth/Weth.tsx:84]: This file doesn't look like its merged in latest with master, its prob worth rebasing now and fix some conflicts

**Comment by nicholaspai** [comment]
@chrismaree also your images of wallet provider displays don't work can you reattach?

File: containers/Connection.ts (hunk 2 above)
Comment (chrismaree) [reply containers/Connection.ts:28]: well this is why it's optional and there are defaults :) if you don't provide them the app still works fine but ideally, you want to have them.

Onboard.js, if provided with your own key, is pretty useful. it gives metrics of the users that connect their wallets and statistics around them.

Comment (chrismaree) [reply containers/Connection.ts:28]: I do agree though that this is the main downside with onboard.js. If we feel really strongly about this we can try a different provider, like [web3modal](https://github.com/Web3Modal/web3modal). I do think though that some providers will _always_ need some kind of keys, such as portis/fortmatic/ImToken.

File: containers/Collateral.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrismaree) [reply containers/Collateral.ts:58]: I _think_ it handles this gracefully. I have not tested this specific case.

File: containers/Connection.ts (hunk 8 above)
Comment (chrismaree) [reply containers/Connection.ts:71]: ye, I'm token is _really_ hard to work with. their `https://eth-testnet.tokenlon.im` is actually just their own kovan node. it's pretty inconsistent and getting it to work has been a major pain in other projects I've worked on before; we could _never_ get imtoken to connect to anything other than their own providers.

This is not a big issue for us given we are on kovan :)

File: containers/Connection.ts (hunk 9 above)
Comment (chrismaree) [reply containers/Connection.ts:79]: ye, it's a bit of a pain but they all need you to give the provider 🤷

**Comment by chrismaree** [comment]
> @chrismaree also your images of wallet provider displays don't work can you reattach?

@nicholaspai do you mean the wallet screenshots in the PR? they load fine for me but here is another screenshot for completeness 
image

image

File: features/weth/Weth.tsx (hunk 10 above)
Comment (chrismaree) [reply features/weth/Weth.tsx:81]: done and updated PR to use master

File: containers/Connection.ts (hunk 2 above)
Comment (nicholaspai) [reply containers/Connection.ts:28]: Ok I'm good with this and using our own onboard.js key since we can get metrics!

However, do you know how we set configs in Vercel?

File: containers/Connection.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicholaspai) [reply containers/Connection.ts:91]: I'd rather just remove it entirely instead of commenting it out

File: containers/Connection.ts (hunk 4 above)
Comment (nicholaspai) [reply containers/Connection.ts:76]: OK well if you see it then that's great. I have a metamask extension on chrome and only see metamask option

File: features/manage-position/Create.tsx (hunk 6 above)
Comment (nicholaspai) [reply features/manage-position/Create.tsx:151]: Ok we can strip out in another PR this one is chunky already

File: containers/Connection.ts [hunk 11]
```diff
@@ -1,54 +1,116 @@
...
   const attemptConnection = async () => {
-    if (window.ethereum === undefined) {
-      throw Error("MetaMask not found, please visit https://metamask.io/");
-    }
-
-    // // for testing
-    // const provider = new ethers.providers.JsonRpcProvider()
-    // let signer = new ethers.Wallet(process.env.PRIV_KEY)
-    // signer = signer.connect(provider)
+    const apiKey = process.env.REACT_APP_ONBOARD_API_KEY
```
Comment (nicholaspai) [inline containers/Connection.ts:28]: For better environment maangement, I'd like to see all `process.env.VAR` refactored into their own `config.js` file. From experience, it's much easier if we have one dedicated file to detecting environment variables and/or setting default values. This way in downstream files such as this, we don't need the `VAR ? VAR : DEFAULT` pattern and can just call `VAR`.

WDYT?

File: containers/Connection.ts (hunk 2 above)
Comment (chrismaree) [reply containers/Connection.ts:28]: Great. It's pretty [easy](https://vercel.com/blog/environment-variables-ui#:~:text=New%20Project%20Setting,field%20in%20the%20Project%20Settings.) to do env vars in Vercel. I'll set them up now before merging this in.

File: features/manage-position/Create.tsx (hunk 6 above)
Comment (chrismaree) [reply features/manage-position/Create.tsx:151]: It covers all functions already though. I'm happy to pull it into two PRs so we can test. 

Another option to "test in production" would be to circulate this testing URL internally for anyone creating a transaction and validate that no errors are thrown.

File: containers/Config.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  const infuraId =
+    process.env.REACT_APP_INFURA_ID || "d5e29c9b9a9d4116a7348113f57770a8";
+  const infuraRpc = `https://${network?.name}.infura.io/v3/${infuraId}`;
+  const apiKey = process.env.REACT_APP_ONBOARD_API_KEY
+    ? process.env.REACT_APP_ONBOARD_API_KEY
+    : "12153f55-f29e-4f11-aa07-90f10da5d778";
+  return {
+    apiKey: apiKey ? apiKey : "12153f55-f29e-4f11-aa07-90f10da5d778",
+    portisKey: process.env.REACT_APP_PORTIS_API_KEY,
+    onboardWalletSelect: {
```
Comment (nicholaspai) [inline containers/Config.ts:13]: nit: I'd recommend you put the `apiKey`, `portisKey`, `onboardWalletSelect` and `wallets` into a `portisConfig` key. I think we'll want to use this container for other external SDK's in the future and the name `apiKey` will be reused a lot

File: containers/Config.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
+import { ethers } from "ethers";
+type Network = ethers.providers.Network;
+export const config = (network: Network | null) => {
```
Comment (nicholaspai) [inline containers/Config.ts:3]: I like passing in `network` here because you could imagine all of the config changing with a diff network

File: containers/Connection.ts [hunk 14]
```diff
@@ -1,54 +1,73 @@
...
 import { useState, useEffect } from "react";
 import { ethers } from "ethers";
+import { API as OnboardApi, Wallet } from "bnc-onboard/dist/src/interfaces";
+import { API as NotifyApi } from "bnc-notify/dist/src/interfaces";
+import Onboard from "bnc-onboard";
+import Notify from "bnc-notify";
 import { Observable } from "rxjs";
 import { debounceTime } from "rxjs/operators";
 
+import { config } from "./Config";
```
Comment (nicholaspai) [inline containers/Connection.ts:11]: If you wrap all the onboard related stuff into an `onboard` key you can then do `const { onboard } = config`,

and then easily access `dappId: onboard.apiKey`

File: containers/Config.ts (hunk 12 above)
Comment (nicholaspai) [reply containers/Config.ts:13]: Another nit: Should these all be CAPITALIZEd variable names since they're all hardcoded values?

i.e.

API_KEY, PORTIS_KEY, ONBOARD_WALLET_SELECT...

File: containers/Config.ts (hunk 13 above)
Comment (chrismaree) [reply containers/Config.ts:3]: ye, I was thinking we could re-use this file for addresses even for network switching (I know there is `EMPs.ts` but this could join all configs in one place for ease of access.

File: features/manage-position/Create.tsx (hunk 6 above)
Comment (chrismaree) [reply features/manage-position/Create.tsx]: I have removed notify.js to put into a separate PR. Cleaner this way.

File: containers/Config.ts (hunk 12 above)
Comment (chrismaree) [reply containers/Config.ts]: I have refactored this accordingly. PTAL. 

I'm not a fan of the all caps on the hardcoded vars though. Also this is a function that can return different valued based on the context. [Airbnb](https://github.com/airbnb/javascript#naming--uppercase) standard says that uppercase constants should only be used if `the programmer can trust it (and its nested properties) to never change` but this is not really the case as the config changes with network param.

File: containers/Connection.ts (hunk 14 above)
Comment (chrismaree) [reply containers/Connection.ts:9]: I don't think the `onboard` object should go in the config. it's more part of `Connection.ts`'s context, where it is right now. 

If we were to move it then it should perhaps go to a `onboard.ts` file that `Connection.ts` could import. This `onboard.js` would also use the `config.ts` file. However, I think this is an over-engineered solution and leaving it where it is now in `Connection.ts` is simple enough while still having sufficient separation of concerns.

**Comment by nicholaspai** [comment]
Actually I agree. Makes sense!

File: containers/Connection.ts [hunk 15]
```diff
@@ -1,54 +1,64 @@
...
-    // let signer = new ethers.Wallet(process.env.PRIV_KEY)
-    // signer = signer.connect(provider)
-
-    // get provider and signer
-    const provider = new ethers.providers.Web3Provider(
-      window.ethereum as ExternalProvider
-    );
-    const signer = provider.getSigner();
-    const network = await provider.getNetwork();
+    const onboard = Onboard({
```
Comment (nicholaspai) [inline containers/Connection.ts:27]: Wait, why do you declare `onboard` here but also in the `useState()` on line 18?

File: containers/Connection.ts [hunk 16]
```diff
@@ -77,7 +87,18 @@ function useConnection() {
     }
   }, [provider]);
 
-  return { provider, signer, network, address, connect, error, block$ };
+  return {
+    provider,
+    onboard,
+
```
Comment (nicholaspai) [inline containers/Connection.ts:93]: Random space here?

File: containers/Connection.ts (hunk 15 above)
Comment (chrismaree) [reply containers/Connection.ts:27]: the reason is that I need to full set up the onboard instance with `await onboardInstance.walletSelect();` before setting into the container hook. I have re-named this variable to make it a bit clear that this is the `onboardInstance`..
